### ISSUE TITLE
what is the actual mean of "x3d" in the output of the epropnp_det?

### ISSUE THREAD

**Issue by andersonhusky** [tjiiv-cprg/EPro-PnP#27]
I am confused about the meaning of "x3d", it seems like all the "x3d" of an object are in one plane, how can PnP calculate pose with this 3d points？
here is my test result about x3d in bev (the first picture)
![demo_x3d_bev](https://user-images.githubusercontent.com/45249955/183364322-a95e3559-9a98-4053-83ff-d62d54cf6663.png)
![demo_3d](https://user-images.githubusercontent.com/45249955/183365024-b091cc7a-3f58-4d76-b58d-ea3b3a0c1ea4.jpg)
![demo_bev](https://user-images.githubusercontent.com/45249955/183365070-1a399625-2781-4446-ad0b-3b4b0db67e08.png)

**Comment by Lakonik** [comment]
Hi! x3d are the points' 3D coordinates in the object coordinate system. In the LineMOD 6DoF pose estimation model, a coordinate regression loss is imposed to force those points to lie on the exact surface of the objects. In the nuScenes 3D detection model though, there is no such loss function and the points are learned to solely minimize the pose loss, so they may not be a clear representation of the object shape.

PS: Your visualizations do seem a bit weird, and I suspect there is something wrong in the model configuration or the visualization code, e.g., the camera intrinsic could be wrong so all the boxes are somewhat tilted in the second image.

**Comment by qinyq** [comment]
> Hi! x3d are the points' 3D coordinates in the object coordinate system. In the LineMOD 6DoF pose estimation model, a coordinate regression loss is imposed to force those points to lie on the exact surface of the objects. In the nuScenes 3D detection model though, there is no such loss function and the points are learned to solely minimize the pose loss, so they may not be a clear representation of the object shape.
> 
> PS: Your visualizations do seem a bit weird, and I suspect there is something wrong in the model configuration or the visualization code, e.g., the camera intrinsic could be wrong so all the boxes are somewhat tilted in the second image.

I also noticed that in the nuScenes 3D model, the x3d points `may not be a clear representation of the object shape`. Is it possible to impose a similar loss func to force those points lie on or nearly on the exact surface of the objects?

Another question: it seems that in EPro-PnP-Det model, there's no label information in the inferred results...right? there are only `['bbox_results', 'bbox_3d_results', 'pose_samples', 'pose_sample_weights']` ...

**Comment by Lakonik** [comment]
> I also noticed that in the nuScenes 3D model, the x3d points `may not be a clear representation of the object shape`. Is it possible to impose a similar loss func to force those points lie on or nearly on the exact surface of the objects?
> 
> Another question: it seems that in EPro-PnP-Det model, there's no label information in the inferred results...right? there are only `['bbox_results', 'bbox_3d_results', 'pose_samples', 'pose_sample_weights']` ...

Yes you can impose such a loss function, e.g., using the LiDAR points as supervisory signal. In EPro-PnP-Det though, we moved the LiDAR-based coordinate regression loss to an auxiliary branch (see supplementary) which is easier to implement. I'm not quite sure how the accuracy would be affected if the loss is imposed on x3d directly.

**Comment by qinyq** [comment]
> Yes you can impose such a loss function, e.g., using the LiDAR points as supervisory signal. In EPro-PnP-Det though, we moved the LiDAR-based coordinate regression loss to an auxiliary branch (see supplementary) which is easier to implement. I'm not quite sure how the accuracy would be affected if the loss is imposed on x3d directly.


Thank you for timely reply! I'll check that branch.
it seems that the official code does not output 'labels' right? I checked the inference result which only has four keys: `['bbox_results', 'bbox_3d_results', 'pose_samples', 'pose_sample_weights']`

**Comment by Lakonik** [comment]
> Thank you for timely reply! I'll check that branch. it seems that the official code does not output 'labels' right? I checked the inference result which only has four keys: `['bbox_results', 'bbox_3d_results', 'pose_samples', 'pose_sample_weights']`

The value corresponding to each key is a list of class-wise results, where the index of each element corresponds to a specific class.

**Comment by qinyq** [comment]
> > Thank you for timely reply! I'll check that branch. it seems that the official code does not output 'labels' right? I checked the inference result which only has four keys: `['bbox_results', 'bbox_3d_results', 'pose_samples', 'pose_sample_weights']`
> 
> The value corresponding to each key is a list of class-wise results, where the index of each element corresponds to a specific class.

aha, I see, thank you!
